PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc

```diff
@@ -112,22 +397,92 @@ void pat::PATIsolatedTrackProducer::produce(edm::StreamID stream, edm::Event& iE
+
+        // shape selection for strips
+        if(isStrip && !DeDxTools::shapeSelection(*(hitInfo->stripCluster(ih))))
+            continue;
         
-        pat::PFIsolation isolationDR03(chiso, nhiso, phiso, puiso);
-        pat::PFIsolation miniIso(chmiso, nhmiso, phmiso, pumiso);
-
-        if(isolationDR03.chargedHadronIso() < absIso_cut ||
-           isolationDR03.chargedHadronIso()/pf_it->p4().pt() < relIso_cut ||
-           miniIso.chargedHadronIso()/pf_it->p4().pt() < miniRelIso_cut){
-            outPtrP->push_back(pat::IsolatedTrack(isolationDR03, miniIso, pf_it->p4(), 
-                                                  pf_it->charge(), pf_it->pdgId(), pf_it->dz(),
-                                                  pf_it->dxy(), pcref));
-        }
+        float Norm=0;       
+        if(isPixel)
+            Norm = 3.61e-06; //compute the normalization factor to get the energy in MeV/mm
+        if(isStrip)
+            Norm = 3.61e-06 * 265;
+            
+        //save the dE/dx in MeV/mm to a vector.  
+        charge_vec.push_back(Norm*hitInfo->charge(ih)/hitInfo->pathlength(ih));           
+    }
 
+    int size = charge_vec.size();
+    float result = 0.0;
+
+    //build the harmonic 2 dE/dx estimator
+    float expo = -2;
+    for(int i=0; i<size; i++){
+        result += pow(charge_vec[i], expo);
     }
+    result = (size>0) ? pow(result/size, 1./expo) : 0.0;
 
+    return result;
+}
+
+TrackDetMatchInfo pat::PATIsolatedTrackProducer::getTrackDetMatchInfo(const edm::Event& iEvent, 
+                                                                      const edm::EventSetup& iSetup,
+                                                                      const reco::Track& track)
+{
+    edm::ESHandle<MagneticField> bField;
+    iSetup.get<IdealMagneticFieldRecord>().get(bField);
+    FreeTrajectoryState initialState = trajectoryStateTransform::initialFreeState(track,&*bField);
+
+    // can't use the associate() using reco::Track directly, since 
+    // track->extra() is non-null but segfaults when trying to use it
+    return trackAssociator_.associate(iEvent, iSetup, trackAssocParameters_, &initialState);
+}
+
+void pat::PATIsolatedTrackProducer::getCaloJetEnergy(const LorentzVector& p4, const reco::CaloJetCollection* cJets,
+                                                     float &caloJetEm, float& caloJetHad) const
+{
+    float nearestDR = 999;
+    int ind = -1;
+    for(unsigned int i=0; i<cJets->size(); i++){
+        float dR = deltaR(cJets->at(i).p4(), p4);
+        if(dR < caloJet_DR_ && dR < nearestDR){
+            nearestDR = dR;
+            ind = i;
+        }
+    }
+
+    if(ind==-1){
+        caloJetEm = 0;
+        caloJetHad = 0;
+    }else{
+        caloJetEm = cJets->at(ind).emEnergyInEB() + cJets->at(ind).emEnergyInEE() + cJets->at(ind).emEnergyInHF();
+        caloJetHad = cJets->at(ind).hadEnergyInHB() + cJets->at(ind).hadEnergyInHE() + cJets->at(ind).hadEnergyInHF();
```

Comment:
```auto const& cJet = cJets->at(ind); ``` to reduce repeated dereference and retrieval calls (even though compiler will likely optimize it out)